File: app/src/main/java/com/example/opengl/Shader.java
package com.example.opengl;

import static android.opengl.GLES20.glUniform1fv;
import static android.opengl.GLES20.glUniform1iv;
import static android.opengl.GLES30.GL_ACTIVE_ATTRIBUTES;
import static android.opengl.GLES30.GL_ACTIVE_UNIFORMS;
import static android.opengl.GLES30.GL_COMPILE_STATUS;
import static android.opengl.GLES30.GL_FRAGMENT_SHADER;
import static android.opengl.GLES30.GL_LINK_STATUS;
import static android.opengl.GLES30.GL_TRUE;
import static android.opengl.GLES30.GL_VERTEX_SHADER;
import static android.opengl.GLES30.glAttachShader;
import static android.opengl.GLES30.glCompileShader;
import static android.opengl.GLES30.glCreateProgram;
import static android.opengl.GLES30.glCreateShader;
import static android.opengl.GLES30.glDeleteShader;
import static android.opengl.GLES30.glDetachShader;
import static android.opengl.GLES30.glGetActiveAttrib;
import static android.opengl.GLES30.glGetActiveUniform;
import static android.opengl.GLES30.glGetAttribLocation;
import static android.opengl.GLES30.glGetProgramInfoLog;
import static android.opengl.GLES30.glGetProgramiv;
import static android.opengl.GLES30.glGetShaderInfoLog;
import static android.opengl.GLES30.glGetShaderiv;
import static android.opengl.GLES30.glGetUniformLocation;
import static android.opengl.GLES30.glLinkProgram;
import static android.opengl.GLES30.glShaderSource;
import static android.opengl.GLES30.glUniform1f;
import static android.opengl.GLES30.glUniform1i;
import static android.opengl.GLES30.glUniform3fv;
import static android.opengl.GLES30.glUniformMatrix4fv;
import static android.opengl.GLES30.glUseProgram;

import android.content.Context;
import android.content.res.Resources;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.IntBuffer;
import java.util.HashMap;
import java.util.Map;

public class Shader {
    public final int Handle;
    private final Map<String, Integer> uniformLocations = new HashMap<>();
    private final Map<String, Integer> attribLocations = new HashMap<>();

    public Shader(Context context, String vertexShaderFilename, String fragmentShaderFilename) throws IOException {
        String shaderSource = readTextFromAssets(context, vertexShaderFilename);
        if (shaderSource.equals(""))
            throw new IOException("Shader file " + vertexShaderFilename + " not valid");

        int vertexShader = glCreateShader(GL_VERTEX_SHADER);
        glShaderSource(vertexShader, shaderSource);
        compileShader(context, vertexShader);

        shaderSource = readTextFromAssets(context, fragmentShaderFilename);
        if (shaderSource.equals(""))
            throw new IOException("Shader file " + fragmentShaderFilename + " not valid");

        int fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
        glShaderSource(fragmentShader, shaderSource);
        compileShader(context, fragmentShader);

        Handle = glCreateProgram();

        glAttachShader(Handle, vertexShader);
        glAttachShader(Handle, fragmentShader);

        linkProgram(context, Handle);

        glDetachShader(Handle, vertexShader);
        glDetachShader(Handle, fragmentShader);
        glDeleteShader(fragmentShader);
        glDeleteShader(vertexShader);

        IntBuffer params = IntBuffer.allocate(1);
        IntBuffer type = IntBuffer.allocate(1);
        glGetProgramiv(Handle, GL_ACTIVE_UNIFORMS, params);
        int numberOfUniforms = params.get(0);

        for (int i = 0; i < numberOfUniforms; i++) {
            String key = glGetActiveUniform(Handle, i, params, type);
            int location = glGetUniformLocation(Handle, key);
            uniformLocations.put(key, location);
        }

        glGetProgramiv(Handle, GL_ACTIVE_ATTRIBUTES, params);
        int numberOfAttributes = params.get(0);
        for (int i = 0; i < numberOfAttributes; i++) {
            String key = glGetActiveAttrib(Handle, i, params, type);
            int location = glGetAttribLocation(Handle, key);
            attribLocations.put(key, location);
        }
    }

    private static void compileShader(Context context, int shader) {
        glCompileShader(shader);

        // Check for compilation errors
        IntBuffer params = IntBuffer.allocate(1);
        glGetShaderiv(shader, GL_COMPILE_STATUS, params);
        if (params.get(0) != GL_TRUE) {
            String infoLog = glGetShaderInfoLog(shader);
            new Handler(Looper.getMainLooper()).post(new Runnable() {
                public void run() {
                    Toast.makeText(context, "Error occurred whilst compiling Shader(" + shader + ").\n\n" + infoLog, Toast.LENGTH_LONG).show();
                }
            });
            //throw new RuntimeException("Error occurred whilst compiling Shader(" + shader + ").\n\n" + infoLog);
        }
    }

    private static void linkProgram(Context context, int program) {
        glLinkProgram(program);

        // Check for linking errors
        IntBuffer params = IntBuffer.allocate(1);
        glGetProgramiv(program, GL_LINK_STATUS, params);
        if (params.get(0) != GL_TRUE) {
            String infoLog = glGetProgramInfoLog(program);
            new Handler(Looper.getMainLooper()).post(new Runnable() {
                public void run() {
                    Toast.makeText(context, "Error occurred whilst linking Program(" + program + ").\n\n" + infoLog, Toast.LENGTH_LONG).show();
                }
            });
            //throw new RuntimeException("Error occurred whilst linking Program(" + program + ").\n\n" + infoLog);
        }
    }

    public void use() {
        glUseProgram(Handle);
    }

    public int getAttribLocation(String attribName) {
        //return GL.GetAttribLocation(Handle, attribName);
        Integer result = attribLocations.get(attribName);
        return result != null ? result : 0;
    }

    public int getUniformLocation(String attribName) {
        Integer result = uniformLocations.get(attribName);
        return result != null ? result : 0;
    }

    // Uniform setters
    public void setInt(String name, int data) {
        glUseProgram(Handle);
        glUniform1i(getUniformLocation(name), data);
    }

    public void setIntArray(String name, int[] data) {
        glUseProgram(Handle);
        glUniform1iv(getUniformLocation(name), data.length, data,0);
    }

    public void setFloat(String name, float data) {
        glUseProgram(Handle);
        glUniform1f(getUniformLocation(name), data);
    }

    public void setFloatArray(String name, float[] data) {
        glUseProgram(Handle);
        glUniform1fv(getUniformLocation(name), data.length, data, 0);
    }

    public void setMatrix4(String name, float[] data) {
        glUseProgram(Handle);
        glUniformMatrix4fv(getUniformLocation(name), 1, false, data, 0); // false - not transpose matrix
    }

    public void setVector3(String name, float[] data) {
        glUseProgram(Handle);
        glUniform3fv(getUniformLocation(name), 1, data, 0);
    }

    public static String readTextFromRaw(Context context, int resourceId) {
        StringBuilder stringBuilder = new StringBuilder();
        try {
            BufferedReader bufferedReader = null;
            try {
                InputStream inputStream = context.getResources().openRawResource(resourceId);
                bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    stringBuilder.append(line);
                    stringBuilder.append("\r\n");
                }
            } finally {
                if (bufferedReader != null) {
                    bufferedReader.close();
                }
            }
        } catch (IOException | Resources.NotFoundException e) {
            e.printStackTrace();
        }
        return stringBuilder.toString();
    }

    public static String readTextFromAssets(Context context, String filepath) {
        StringBuilder stringBuilder = new StringBuilder();
        try {
            BufferedReader bufferedReader = null;
            try {
                InputStream inputStream = context.getAssets().open(filepath);
                bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    stringBuilder.append(line);
                    stringBuilder.append("\r\n");
                }
            } finally {
                if (bufferedReader != null) {
                    bufferedReader.close();
                }
            }
        } catch (IOException e) {
            new Handler(Looper.getMainLooper()).post(new Runnable() {
                public void run() {
                    Toast.makeText(context, "Error occurred whilst loading Shader(" + filepath + ").\n\n" + e.getMessage(), Toast.LENGTH_LONG).show();
                }
            });
        }
        return stringBuilder.toString();
    }
}

File: app/src/main/java/com/example/opengl/Shader.kt
package com.example.opengl

import android.content.Context
import android.content.res.Resources.NotFoundException
import android.opengl.GLES20
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.nio.IntBuffer

class Shader(context: Context, vertexShaderFilename: String, fragmentShaderFilename: String) {
    private val handle: Int
    private val uniformLocations: MutableMap<String, Int> = HashMap()
    private val attribLocations: MutableMap<String, Int> = HashMap()

    init {
        var shaderSource = readTextFromAssets(context, vertexShaderFilename)
        if (shaderSource == "") throw IOException("Shader file $vertexShaderFilename not valid")
        val vertexShader = GLES20.glCreateShader(GLES20.GL_VERTEX_SHADER)
        GLES20.glShaderSource(vertexShader, shaderSource)
        compileShader(context, vertexShader)
        shaderSource = readTextFromAssets(context, fragmentShaderFilename)
        if (shaderSource == "") throw IOException("Shader file $fragmentShaderFilename not valid")
        val fragmentShader = GLES20.glCreateShader(GLES20.GL_FRAGMENT_SHADER)
        GLES20.glShaderSource(fragmentShader, shaderSource)
        compileShader(context, fragmentShader)
        handle = GLES20.glCreateProgram()
        GLES20.glAttachShader(handle, vertexShader)
        GLES20.glAttachShader(handle, fragmentShader)
        linkProgram(context, handle)
        GLES20.glDetachShader(handle, vertexShader)
        GLES20.glDetachShader(handle, fragmentShader)
        GLES20.glDeleteShader(fragmentShader)
        GLES20.glDeleteShader(vertexShader)
        val params = IntBuffer.allocate(1)
        val type = IntBuffer.allocate(1)
        GLES20.glGetProgramiv(handle, GLES20.GL_ACTIVE_UNIFORMS, params)
        val numberOfUniforms = params[0]
        for (i in 0 until numberOfUniforms) {
            val key = GLES20.glGetActiveUniform(handle, i, params, type)
            val location = GLES20.glGetUniformLocation(handle, key)
            uniformLocations[key] = location
        }
        GLES20.glGetProgramiv(handle, GLES20.GL_ACTIVE_ATTRIBUTES, params)
        val numberOfAttributes = params[0]
        for (i in 0 until numberOfAttributes) {
            val key = GLES20.glGetActiveAttrib(handle, i, params, type)
            val location = GLES20.glGetAttribLocation(handle, key)
            attribLocations[key] = location
        }
    }

    fun use() {
        GLES20.glUseProgram(handle)
    }

    fun getAttribLocation(attribName: String): Int {
        //return GL.GetAttribLocation(Handle, attribName);
        val result = attribLocations[attribName]
        return result ?: 0
    }

    fun getUniformLocation(attribName: String): Int {
        val result = uniformLocations[attribName]
        return result ?: 0
    }

    // Uniform setters
    fun setInt(name: String, data: Int) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform1i(getUniformLocation(name), data)
    }

    fun setIntArray(name: String, data: IntArray) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform1iv(getUniformLocation(name), data.size, data, 0)
    }

    fun setFloat(name: String, data: Float) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform1f(getUniformLocation(name), data)
    }

    fun setFloatArray(name: String, data: FloatArray) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform1fv(getUniformLocation(name), data.size, data, 0)
    }

    fun setMatrix4(name: String, data: FloatArray?) {
        GLES20.glUseProgram(handle)
        GLES20.glUniformMatrix4fv(
            getUniformLocation(name),
            1,
            false,
            data,
            0
        ) // false - not transpose matrix
    }

    fun setVector3(name: String, data: FloatArray?) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform3fv(getUniformLocation(name), 1, data, 0)
    }

    fun setVector3(name: String, x: Float, y: Float, z: Float) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform3f(getUniformLocation(name), x, y, z)
    }

    fun setVector2(name: String, data: FloatArray?) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform2fv(getUniformLocation(name), 1, data, 0)
    }

    fun setVector2(name: String, x: Float, y: Float) {
        GLES20.glUseProgram(handle)
        GLES20.glUniform2f(getUniformLocation(name), x, y)
    }

    companion object {
        private fun compileShader(context: Context, shader: Int) {
            GLES20.glCompileShader(shader)

            // Check for compilation errors
            val params = IntBuffer.allocate(1)
            GLES20.glGetShaderiv(shader, GLES20.GL_COMPILE_STATUS, params)
            if (params[0] != GLES20.GL_TRUE) {
                val infoLog = GLES20.glGetShaderInfoLog(shader)
                Handler(Looper.getMainLooper()).post {
                    Toast.makeText(
                        context,
                        "Error occurred whilst compiling Shader($shader).\n\n$infoLog",
                        Toast.LENGTH_LONG
                    ).show()
                }
                //throw new RuntimeException("Error occurred whilst compiling Shader(" + shader + ").\n\n" + infoLog);
            }
        }

        private fun linkProgram(context: Context, program: Int) {
            GLES20.glLinkProgram(program)

            // Check for linking errors
            val params = IntBuffer.allocate(1)
            GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, params)
            if (params[0] != GLES20.GL_TRUE) {
                val infoLog = GLES20.glGetProgramInfoLog(program)
                Handler(Looper.getMainLooper()).post {
                    Toast.makeText(
                        context,
                        "Error occurred whilst linking Program($program).\n\n$infoLog",
                        Toast.LENGTH_LONG
                    ).show()
                }
                //throw new RuntimeException("Error occurred whilst linking Program(" + program + ").\n\n" + infoLog);
            }
        }

        fun readTextFromRaw(context: Context, resourceId: Int): String {
            val stringBuilder = StringBuilder()
            try {
                var bufferedReader: BufferedReader? = null
                try {
                    val inputStream = context.resources.openRawResource(resourceId)
                    bufferedReader = BufferedReader(InputStreamReader(inputStream))
                    var line: String?
                    while (bufferedReader.readLine().also { line = it } != null) {
                        stringBuilder.append(line)
                        stringBuilder.append("\r\n")
                    }
                } finally {
                    bufferedReader?.close()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: NotFoundException) {
                e.printStackTrace()
            }
            return stringBuilder.toString()
        }

        fun readTextFromAssets(context: Context, filepath: String): String {
            val stringBuilder = StringBuilder()
            try {
                var bufferedReader: BufferedReader? = null
                try {
                    val inputStream = context.assets.open(filepath)
                    bufferedReader = BufferedReader(InputStreamReader(inputStream))
                    var line: String?
                    while (bufferedReader.readLine().also { line = it } != null) {
                        stringBuilder.append(line)
                        stringBuilder.append("\r\n")
                    }
                } finally {
                    bufferedReader?.close()
                }
            } catch (e: IOException) {
                Handler(Looper.getMainLooper()).post {
                    Toast.makeText(
                        context, """
     Error occurred whilst loading Shader($filepath).    
     ${e.message}
     """.trimIndent(), Toast.LENGTH_LONG
                    ).show()
                }
            }
            return stringBuilder.toString()
        }
    }
}